ldms/src/core/ldms_log.c

```diff
@@ -0,0 +1,339 @@
+{
+	int level = -1;
+	if (0 == strcmp(verbose_level, "QUIET")) {
+		quiet = 1;
+		level = LDMSD_LLASTLEVEL;
+	} else {
+		level = ldms_str_to_loglevel(verbose_level);
+		quiet = 0;
+	}
+	if (level < 0)
+		return level;
+	log_level_thr = level;
+	return 0;
+}
+
+enum ldms_loglevel ldms_loglevel_get()
+{
+	return log_level_thr;
+}
+
+int ldms_loglevel_to_syslog(enum ldms_loglevel level)
+{
+	switch (level) {
+#define MAPLOG(X,Y) case LDMSD_L##X: return LOG_##Y
+	MAPLOG(DEBUG,DEBUG);
+	MAPLOG(INFO,INFO);
+	MAPLOG(WARNING,WARNING);
+	MAPLOG(ERROR,ERR);
+	MAPLOG(CRITICAL,CRIT);
+	MAPLOG(ALL,ALERT);
+	default:
+		return LOG_ERR;
+	}
+#undef MAPLOG
+}
+
+/* Impossible file pointer as syslog-use sentinel */
+#define LDMSD_LOG_SYSLOG ((FILE*)0x7)
+
+void __ldms_log(enum ldms_loglevel level, const char *fmt, va_list ap)
+{
+	if ((level != LDMSD_LALL) &&
+			(quiet || ((0 <= level) && (level < log_level_thr))))
+		return;
+	if (log_fp == LDMSD_LOG_SYSLOG) {
+		vsyslog(ldms_loglevel_to_syslog(level),fmt,ap);
+		return;
+	}
+	time_t t;
+	struct tm tm;
+	char dtsz[200];
+
+	pthread_mutex_lock(&log_lock);
+	if (!log_fp) {
+		pthread_mutex_unlock(&log_lock);
+		return;
+	}
+	t = time(NULL);
+	localtime_r(&t, &tm);
+	if (strftime(dtsz, sizeof(dtsz), "%a %b %d %H:%M:%S %Y", &tm))
+		fprintf(log_fp, "%s: ", dtsz);
+
+	if (level < LDMSD_LALL) {
+		fprintf(log_fp, "%-10s: ", ldms_loglevel_names[level]);
+	}
+
+	vfprintf(log_fp, fmt, ap);
+	fflush(log_fp);
+	pthread_mutex_unlock(&log_lock);
+}
+
+void ldms_log(enum ldms_loglevel level, const char *fmt, ...)
+{
+	va_list ap;
+	va_start(ap, fmt);
+	__ldms_log(level, fmt, ap);
+	va_end(ap);
+}
+
+/* All messages from the ldms library are of e level.*/
+#define LDMSD_LOG_AT(e,fsuf) \
+void ldms_l##fsuf(const char *fmt, ...) \
+{ \
+	va_list ap; \
+	va_start(ap, fmt); \
+	__ldms_log(e, fmt, ap); \
+	va_end(ap); \
+}
+
+LDMSD_LOG_AT(LDMSD_LDEBUG,debug);
+LDMSD_LOG_AT(LDMSD_LINFO,info);
+LDMSD_LOG_AT(LDMSD_LWARNING,warning);
+LDMSD_LOG_AT(LDMSD_LERROR,error);
+LDMSD_LOG_AT(LDMSD_LCRITICAL,critical);
+LDMSD_LOG_AT(LDMSD_LALL,all);
+
+static char msg_buf[4096];
+void ldms_msg_logger(enum ldms_loglevel level, const char *fmt, ...)
+{
+	va_list ap;
+	va_start(ap, fmt);
+	vsnprintf(msg_buf, sizeof(msg_buf), fmt, ap);
+	ldms_log(level, "%s", msg_buf);
+}
+
+enum ldms_loglevel ldms_str_to_loglevel(const char *level_s)
+{
+	int i;
+	for (i = 0; i < LDMSD_LLASTLEVEL; i++)
+		if (0 == strcasecmp(level_s, ldms_loglevel_names[i]))
+			return i;
+	if (strcasecmp(level_s,"QUIET") == 0) {
+		return LDMSD_LALL;
+				}
+	if (strcasecmp(level_s,"ALWAYS") == 0) {
+		return LDMSD_LALL;
+	}
+	if (strcasecmp(level_s,"CRIT") == 0) {
+		return LDMSD_LCRITICAL;
+	}
+	return LDMSD_LNONE;
+}
+
+const char *ldms_loglevel_to_str(enum ldms_loglevel level)
+{
+	if ((level >= LDMSD_LDEBUG) && (level < LDMSD_LLASTLEVEL))
+		return ldms_loglevel_names[level];
+	return "LDMSD_LNONE";
+}
+
+/** return a file pointer or a special syslog pointer */
```

Comment:
Nit: this actually sets a global variable and returns an errno.

Reply:
I tried to keep this close to the original code as possible without being broken.  Will take another look.